File: src/lib/integrations/slack.ts
import { WebClient } from "@slack/web-api";
import type { Prisma } from "@/generated/prisma/client";
import { db } from "@/lib/db";
import {
  enqueueIntentExtraction,
  scheduleLiveNotesUpdate,
} from "@/lib/queue";

export async function getSlackClient(userId: string): Promise<WebClient | null> {
  const integration = await db.integration.findUnique({
    where: { userId_provider: { userId, provider: "SLACK" } },
  });

  if (!integration) return null;
  return new WebClient(integration.accessToken);
}

export async function isSlackConnected(userId: string) {
  const integration = await db.integration.findUnique({
    where: { userId_provider: { userId, provider: "SLACK" } },
  });
  return !!integration;
}

export async function getSlackMetadata(userId: string) {
  const integration = await db.integration.findUnique({
    where: { userId_provider: { userId, provider: "SLACK" } },
  });
  if (!integration) return null;
  return integration.metadata as {
    teamId?: string;
    teamName?: string;
    slackUserId?: string;
    autoHuddleCapture?: boolean;
  } | null;
}

export async function updateSlackSettings(
  userId: string,
  settings: { autoHuddleCapture?: boolean }
) {
  const integration = await db.integration.findUnique({
    where: { userId_provider: { userId, provider: "SLACK" } },
  });
  if (!integration) throw new Error("Slack not connected");

  const metadata = (integration.metadata as Record<string, unknown>) ?? {};
  await db.integration.update({
    where: { id: integration.id },
    data: {
      metadata: { ...metadata, ...settings } as Prisma.InputJsonValue,
    },
  });
}

export async function saveSlackIntegration(
  userId: string,
  accessToken: string,
  metadata?: Record<string, unknown>
) {
  await db.integration.upsert({
    where: { userId_provider: { userId, provider: "SLACK" } },
    create: {
      userId,
      provider: "SLACK",
      accessToken,
      metadata: (metadata ?? {}) as Prisma.InputJsonValue,
    },
    update: {
      accessToken,
      metadata: (metadata ?? {}) as Prisma.InputJsonValue,
    },
  });
}

export async function listSlackChannels(userId: string) {
  const client = await getSlackClient(userId);
  if (!client) return [];

  const [channels, ims] = await Promise.all([
    client.conversations.list({ types: "public_channel,private_channel", limit: 50 }),
    client.conversations.list({ types: "im", limit: 50 }),
  ]);

  const items = [
    ...(channels.channels ?? []).map((c) => ({
      id: c.id!,
      name: c.name ?? c.id!,
      type: "channel" as const,
    })),
    ...(ims.channels ?? []).map((c) => ({
      id: c.id!,
      name: c.user ? `DM:${c.user}` : c.id!,
      type: "im" as const,
    })),
  ];

  return items;
}

export async function fetchChannelHistory(
  userId: string,
  channelId: string,
  limit = 50
) {
  const client = await getSlackClient(userId);
  if (!client) return [];

  const result = await client.conversations.history({ channel: channelId, limit });
  const messages = result.messages ?? [];

  const userIds = [...new Set(messages.map((m) => m.user).filter(Boolean))] as string[];
  const users: Record<string, string> = {};

  for (const uid of userIds) {
    try {
      const info = await client.users.info({ user: uid });
      users[uid] = info.user?.real_name ?? info.user?.name ?? uid;
    } catch {
      users[uid] = uid;
    }
  }

  return messages
    .filter((m) => m.text && !m.subtype)
    .reverse()
    .map((m) => ({
      externalId: m.ts!,
      speaker: m.user ? users[m.user] ?? m.user : "Unknown",
      content: m.text!,
      sentAt: new Date(parseFloat(m.ts!) * 1000),
    }));
}

export async function handleSlackMessage(
  channelId: string,
  message: {
    ts: string;
    user?: string;
    text?: string;
  }
) {
  const activeSessions = await db.captureSession.findMany({
    where: {
      status: "ACTIVE",
      sourceType: "SLACK",
      sourceRef: channelId,
    },
  });

  if (activeSessions.length === 0 || !message.text) return;

  for (const session of activeSessions) {
    const existing = await db.message.findFirst({
      where: { sessionId: session.id, externalId: message.ts },
    });
    if (existing) continue;

    let speaker = "Unknown";
    const integration = await db.integration.findUnique({
      where: { userId_provider: { userId: session.userId, provider: "SLACK" } },
    });

    if (integration && message.user) {
      const client = new WebClient(integration.accessToken);
      try {
        const info = await client.users.info({ user: message.user });
        speaker = info.user?.real_name ?? info.user?.name ?? message.user;
      } catch {
        speaker = message.user;
      }
    }

    await db.message.create({
      data: {
        sessionId: session.id,
        externalId: message.ts,
        speaker,
        content: message.text,
        sentAt: new Date(parseFloat(message.ts) * 1000),
      },
    });

    await enqueueIntentExtraction(session.id);
    scheduleLiveNotesUpdate(session.id);
  }
}

export function verifySlackSignature(
  signingSecret: string,
  signature: string,
  timestamp: string,
  body: string
): boolean {
  const crypto = require("crypto");
  const fiveMinutesAgo = Math.floor(Date.now() / 1000) - 60 * 5;
  if (parseInt(timestamp, 10) < fiveMinutesAgo) return false;

  const sigBasestring = `v0:${timestamp}:${body}`;
  const mySignature =
    "v0=" +
    crypto.createHmac("sha256", signingSecret).update(sigBasestring).digest("hex");

  return crypto.timingSafeEqual(
    Buffer.from(mySignature),
    Buffer.from(signature)
  );
}

File: backend/app/services/integrations/slack.py
import hashlib
import hmac
import time
from datetime import datetime, timezone
from typing import Any

from slack_sdk import WebClient
from sqlalchemy import select

from app.core.ids import generate_id

from app.database import AsyncSessionLocal
from app.models import (
    CaptureSession,
    CaptureSessionStatus,
    CaptureSourceType,
    Integration,
    IntegrationProvider,
    Message,
)
from app.queue import enqueue_intent_extraction, schedule_live_notes_update


async def get_slack_client(user_id: str) -> WebClient | None:
    async with AsyncSessionLocal() as db:
        result = await db.execute(
            select(Integration).where(
                Integration.userId == user_id,
                Integration.provider == IntegrationProvider.SLACK,
            )
        )
        integration = result.scalar_one_or_none()
    if not integration:
        return None
    return WebClient(token=integration.accessToken)


async def is_slack_connected(user_id: str) -> bool:
    async with AsyncSessionLocal() as db:
        result = await db.execute(
            select(Integration).where(
                Integration.userId == user_id,
                Integration.provider == IntegrationProvider.SLACK,
            )
        )
        return result.scalar_one_or_none() is not None


async def get_slack_metadata(user_id: str) -> dict[str, Any] | None:
    async with AsyncSessionLocal() as db:
        result = await db.execute(
            select(Integration).where(
                Integration.userId == user_id,
                Integration.provider == IntegrationProvider.SLACK,
            )
        )
        integration = result.scalar_one_or_none()
    return integration.metadata_ if integration else None


async def update_slack_settings(
    user_id: str,
    settings: dict[str, Any],
) -> None:
    async with AsyncSessionLocal() as db:
        result = await db.execute(
            select(Integration).where(
                Integration.userId == user_id,
                Integration.provider == IntegrationProvider.SLACK,
            )
        )
        integration = result.scalar_one_or_none()
        if not integration:
            raise RuntimeError("Slack not connected")

        metadata = dict(integration.metadata_ or {})
        metadata.update(settings)
        integration.metadata_ = metadata
        await db.commit()


async def save_slack_integration(
    user_id: str,
    access_token: str,
    metadata: dict[str, Any] | None = None,
) -> None:
    meta = metadata or {}
    async with AsyncSessionLocal() as db:
        result = await db.execute(
            select(Integration).where(
                Integration.userId == user_id,
                Integration.provider == IntegrationProvider.SLACK,
            )
        )
        integration = result.scalar_one_or_none()

        if integration:
            integration.accessToken = access_token
            integration.metadata_ = meta
        else:
            db.add(
                Integration(
                    id=generate_id(),
                    userId=user_id,
                    provider=IntegrationProvider.SLACK,
                    accessToken=access_token,
                    metadata_=meta,
                )
            )
        await db.commit()


async def list_slack_channels(user_id: str) -> list[dict[str, str]]:
    client = await get_slack_client(user_id)
    if not client:
        return []

    channels_result = client.conversations_list(
        types="public_channel,private_channel", limit=50
    )
    ims_result = client.conversations_list(types="im", limit=50)

    items: list[dict[str, str]] = []
    for c in channels_result.get("channels") or []:
        items.append(
            {
                "id": c["id"],
                "name": c.get("name") or c["id"],
                "type": "channel",
            }
        )
    for c in ims_result.get("channels") or []:
        items.append(
            {
                "id": c["id"],
                "name": f"DM:{c.get('user')}" if c.get("user") else c["id"],
                "type": "im",
            }
        )
    return items


async def fetch_channel_history(
    user_id: str,
    channel_id: str,
    limit: int = 50,
) -> list[dict[str, Any]]:
    client = await get_slack_client(user_id)
    if not client:
        return []

    result = client.conversations_history(channel=channel_id, limit=limit)
    messages = result.get("messages") or []

    user_ids = list({m.get("user") for m in messages if m.get("user")})
    users: dict[str, str] = {}

    for uid in user_ids:
        try:
            info = client.users_info(user=uid)
            user_data = info.get("user") or {}
            users[uid] = user_data.get("real_name") or user_data.get("name") or uid
        except Exception:
            users[uid] = uid

    return [
        {
            "externalId": m["ts"],
            "speaker": users.get(m.get("user"), m.get("user") or "Unknown"),
            "content": m["text"],
            "sentAt": datetime.fromtimestamp(float(m["ts"]), tz=timezone.utc),
        }
        for m in reversed(messages)
        if m.get("text") and not m.get("subtype")
    ]


async def handle_slack_message(
    channel_id: str,
    message: dict[str, Any],
) -> None:
    async with AsyncSessionLocal() as db:
        session_result = await db.execute(
            select(CaptureSession).where(
                CaptureSession.status == CaptureSessionStatus.ACTIVE,
                CaptureSession.sourceType == CaptureSourceType.SLACK,
                CaptureSession.sourceRef == channel_id,
            )
        )
        active_sessions = session_result.scalars().all()

    if not active_sessions or not message.get("text"):
        return

    for session in active_sessions:
        async with AsyncSessionLocal() as db:
            existing_result = await db.execute(
                select(Message).where(
                    Message.sessionId == session.id,
                    Message.externalId == message["ts"],
                )
            )
            if existing_result.scalar_one_or_none():
                continue

            speaker = "Unknown"
            integration_result = await db.execute(
                select(Integration).where(
                    Integration.userId == session.userId,
                    Integration.provider == IntegrationProvider.SLACK,
                )
            )
            integration = integration_result.scalar_one_or_none()

            if integration and message.get("user"):
                client = WebClient(token=integration.accessToken)
                try:
                    info = client.users_info(user=message["user"])
                    user_data = info.get("user") or {}
                    speaker = (
                        user_data.get("real_name")
                        or user_data.get("name")
                        or message["user"]
                    )
                except Exception:
                    speaker = message["user"]

            db.add(
                Message(
                    id=generate_id(),
                    sessionId=session.id,
                    externalId=message["ts"],
                    speaker=speaker,
                    content=message["text"],
                    sentAt=datetime.fromtimestamp(float(message["ts"]), tz=timezone.utc),
                )
            )
            await db.commit()

        await enqueue_intent_extraction(session.id)
        schedule_live_notes_update(session.id)


def verify_slack_signature(
    signing_secret: str,
    signature: str,
    timestamp: str,
    body: str,
) -> bool:
    five_minutes_ago = int(time.time()) - 60 * 5
    if int(timestamp) < five_minutes_ago:
        return False

    sig_basestring = f"v0:{timestamp}:{body}"
    my_signature = (
        "v0="
        + hmac.new(
            signing_secret.encode(),
            sig_basestring.encode(),
            hashlib.sha256,
        ).hexdigest()
    )
    return hmac.compare_digest(my_signature, signature)
